paramiko/pkey.py

```diff
@@ -52,36 +178,121 @@ def unpad(data):
     return data[:-padding_length]
 
 
+def _read_private_key_old_format(headers, data, password):
+    if 'proc-type' not in headers:
+        # unencryped: done
+        return data
+    # encrypted keyfile: will need a password
+    proc_type = headers['proc-type']
+    if proc_type != '4,ENCRYPTED':
+        raise SSHException(
+            'Unknown private key structure "{}"'.format(proc_type)
+        )
+    try:
+        encryption_type, saltstr = headers['dek-info'].split(',')
+    except:
+        raise SSHException("Can't parse DEK-info in private key file")
+    if encryption_type not in _CIPHER_TABLE:
+        raise SSHException(
+            'Unknown private key cipher "{}"'.format(encryption_type))
+    # if no password was passed in,
+    # raise an exception pointing out that we need one
+    if password is None:
+        raise PasswordRequiredException('Private key file is encrypted')
+    cipher = _CIPHER_TABLE[encryption_type]['cipher']
+    keysize = _CIPHER_TABLE[encryption_type]['keysize']
+    mode = _CIPHER_TABLE[encryption_type]['mode']
+    salt = unhexlify(b(saltstr))
+    key = util.generate_key_bytes(md5, salt, password, keysize)
+    decryptor = Cipher(
+        cipher(key), mode(salt), backend=default_backend()
+    ).decryptor()
+    return decryptor.update(data) + decryptor.finalize()
+
+
+def _read_private_key_new_format(data, password):
```

Comment:
The `_read_private_key_old_format` and `_read_private_key_new_format` functions were extracted from the `PKey` class but otherwise changed very little.

Reply:
In fact, I have reverted the extraction of these two functions.

After looking at this with a fresh pair of eyes today, you were right, keeping these two functions in `PKey` isn't that bad and makes the diff **so much easier** to read, that by itself is a significant benefit.

(`_read_private_key` being the execption because I re-wrote that significantly into `_parse_openssh_pkey`)